File: Analisador Lexico/scanner.py
import re
import json
import sys


# Sequências de escape válidas
ESCAPE = r"\\(?:n|t|\\|'|\")"


TOKEN_SPEC = [

    # =========================
    # Palavras reservadas
    # =========================

    ("INT", r"\bint\b"),
    ("FLOAT", r"\bfloat\b"),
    ("BOOL", r"\bbool\b"),
    ("CHAR", r"\bchar\b"),
    ("VOID", r"\bvoid\b"),
    ("IF", r"\bif\b"),
    ("ELSE", r"\belse\b"),
    ("WHILE", r"\bwhile\b"),
    ("FOR", r"\bfor\b"),
    ("RETURN", r"\breturn\b"),
    ("BREAK", r"\bbreak\b"),
    ("CONTINUE", r"\bcontinue\b"),
    ("TRUE", r"\btrue\b"),
    ("FALSE", r"\bfalse\b"),
    ("PRINT", r"\bprint\b"),
    ("READ", r"\bread\b"),


    # =========================
    # Comentários inválidos
    # =========================

    # Comentários válidos PRIMEIRO
    ("COMMENT", r"//[^\n]*"),
    ("BLOCK_COMMENT", r"/\*[\s\S]*?\*/"),

    # Comentário que começa mas não termina
    ("UNTERMINATED_BLOCK_COMMENT", r"/\*[\s\S]*$"),


    # =========================
    # Strings e caracteres
    # =========================

    # Casos válidos
    ("STRING_LIT", rf'"(?:[^"\\]|{ESCAPE})*"'),
    ("CHAR_LIT", rf"'(?:[^'\\]|{ESCAPE})'"),

    # Casos inválidos
    ("UNTERMINATED_STRING", r'"[^"\n)]*'),
    ("UNTERMINATED_CHAR", r"'[^\n]*"),


    # =========================
    # Identificadores e números
    # =========================

    ("IDENT", r"[a-zA-Z_][a-zA-Z0-9_]*"),

    ("FLOAT_LIT", r"[0-9]+\.[0-9]+"),

    ("INT_LIT", r"[0-9]+"),


    # =========================
    # Operadores
    # =========================

    ("EQ", r"=="),
    ("NE", r"!="),
    ("LE", r"<="),
    ("GE", r">="),
    ("AND", r"&&"),
    ("OR", r"\|\|"),

    ("PLUS", r"\+"),
    ("MINUS", r"-"),
    ("STAR", r"\*"),
    ("SLASH", r"/"),
    ("PERCENT", r"%"),

    ("LT", r"<"),
    ("GT", r">"),
    ("NOT", r"!"),
    ("ASSIGN", r"="),


    # =========================
    # Delimitadores
    # =========================

    ("LPAREN", r"\("),
    ("RPAREN", r"\)"),

    ("LBRACE", r"\{"),
    ("RBRACE", r"\}"),

    ("LBRACKET", r"\["),
    ("RBRACKET", r"\]"),

    ("SEMICOLON", r";"),
    ("COMMA", r","),

    # Ponto
    ("DOT", r"\."),


    # =========================
    # Espaços e quebras de linha
    # =========================

    ("WHITESPACE", r"\s+"),
]


# Junta todos os padrões em uma única expressão regular
tok_regex = "|".join(
    f"(?P<{name}>{pattern})"
    for name, pattern in TOKEN_SPEC
)

get_token = re.compile(tok_regex).match


def get_attribute(tipo, valor):

    if tipo == "IDENT":
        return valor

    if tipo == "INT_LIT":
        return int(valor)

    if tipo == "FLOAT_LIT":
        return float(valor)

    if tipo == "STRING_LIT":
        return valor[1:-1]

    if tipo == "CHAR_LIT":
        return valor[1:-1]

    return None


def lexer(codigo):

    pos = 0
    linha = 1
    coluna = 1

    # Guarda se ocorreu algum erro durante a análise
    teve_erro = False


    while pos < len(codigo):

        match = get_token(codigo, pos)


        # =========================
        # Símbolo desconhecido
        # =========================

        if not match:

            erro = {
                "error": "UNKNOWN_SYMBOL",
                "lexeme": codigo[pos],
                "line": linha,
                "column": coluna
            }

            print(json.dumps(erro, ensure_ascii=False))

            teve_erro = True

            # Avança um caractere para evitar loop infinito
            if codigo[pos] == "\n":
                linha += 1
                coluna = 1
            else:
                coluna += 1

            pos += 1

            # Continua analisando o restante do código
            continue


        tipo = match.lastgroup
        valor = match.group()

        linha_inicio = linha
        coluna_inicio = coluna

        pos = match.end()


        # =========================
        # Atualiza linha e coluna
        # =========================

        linhas = valor.split("\n")

        if len(linhas) > 1:

            linha += len(linhas) - 1
            coluna = len(linhas[-1]) + 1

        else:

            coluna += len(valor)


        # =========================
        # Erro: string não terminada
        # =========================

        if tipo == "UNTERMINATED_STRING":

            erro = {
                "error": "UNTERMINATED_STRING",
                "lexeme": valor,
                "line": linha_inicio,
                "column": coluna_inicio
            }

            print(json.dumps(erro, ensure_ascii=False))

            teve_erro = True

            continue


        # =========================
        # Erro: caractere não terminado
        # =========================

        if tipo == "UNTERMINATED_CHAR":

            erro = {
                "error": "UNTERMINATED_CHAR",
                "lexeme": valor,
                "line": linha_inicio,
                "column": coluna_inicio
            }

            print(json.dumps(erro, ensure_ascii=False))

            teve_erro = True

            continue


        # =========================
        # Erro: comentário não terminado
        # =========================

        if tipo == "UNTERMINATED_BLOCK_COMMENT":

            erro = {
                "error": "UNTERMINATED_BLOCK_COMMENT",
                "lexeme": valor,
                "line": linha_inicio,
                "column": coluna_inicio
            }

            print(json.dumps(erro, ensure_ascii=False))

            teve_erro = True

            continue


        # =========================
        # Ignora espaços e comentários
        # =========================

        if tipo in (
            "WHITESPACE",
            "COMMENT",
            "BLOCK_COMMENT"
        ):

            continue


        # =========================
        # Token válido
        # =========================

        attribute = get_attribute(tipo, valor)

        token = {
            "token": tipo,
            "lexeme": valor,
            "attribute": attribute,
            "line": linha_inicio,
            "column": coluna_inicio
        }

        print(json.dumps(token, ensure_ascii=False))


    # =========================
    # EOF
    # =========================

    token = {
        "token": "EOF",
        "lexeme": "",
        "attribute": None,
        "line": linha,
        "column": coluna
    }

    print(json.dumps(token, ensure_ascii=False))


    # Se ocorreu algum erro, retorna 2.
    # Caso contrário, retorna 0.
    if teve_erro:
        return 2

    return 0


def ler_arquivo(nome_arquivo):

    try:

        with open(
            nome_arquivo,
            "r",
            encoding="utf-8"
        ) as arquivo:

            return arquivo.read()

    except FileNotFoundError:

        raise FileNotFoundError(
            f"Arquivo não encontrado: {nome_arquivo}"
        )


def main():

    if len(sys.argv) != 2:

        print("Uso: python scanner.py <arquivo.c>")

        return 1


    nome_arquivo = sys.argv[1]


    try:

        codigo = ler_arquivo(nome_arquivo)

        return lexer(codigo)


    except FileNotFoundError as erro:

        print(erro)

        return 1


if __name__ == "__main__":

    sys.exit(main())

File: Analisador Lexico/scanner.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>
#include <ctype.h>

/* ============================================================
   ESPECIFICAÇÃO DOS TOKENS
   ============================================================ */

typedef struct {
    const char *name;
    const char *pattern;
} TokenSpec;

TokenSpec TOKEN_SPEC[] = {

    {"AUTO",       "auto"},
    {"BREAK",      "break"},
    {"CASE",       "case"},
    {"CHAR",       "char"},
    {"CONST",      "const"},
    {"CONTINUE",   "continue"},
    {"DEFAULT",    "default"},
    {"DO",         "do"},
    {"DOUBLE",     "double"},
    {"ELSE",       "else"},
    {"ENUM",       "enum"},
    {"EXTERN",     "extern"},
    {"FLOAT",      "float"},
    {"FOR",        "for"},
    {"GOTO",       "goto"},
    {"IF",         "if"},
    {"INT",        "int"},
    {"LONG",       "long"},
    {"REGISTER",   "register"},
    {"RETURN",     "return"},
    {"SHORT",      "short"},
    {"SIGNED",     "signed"},
    {"SIZEOF",     "sizeof"},
    {"STATIC",     "static"},
    {"STRUCT",     "struct"},
    {"SWITCH",     "switch"},
    {"TYPEDEF",    "typedef"},
    {"UNION",      "union"},
    {"UNSIGNED",   "unsigned"},
    {"VOID",       "void"},
    {"VOLATILE",   "volatile"},
    {"WHILE",      "while"},
    {"BOOL",       "bool"},
    {"TRUE",       "true"},
    {"FALSE",      "false"},
    {"PRINT",      "print"},

    {"COMMENT",       "//[^\n]*"},
    {"BLOCK_COMMENT", "/\\*[\\s\\S]*\\*/"},

    {"IDENT",      "[a-zA-Z_][a-zA-Z0-9_]*"},

    {"FLOAT_LIT",  "[0-9]+\\.[0-9]+"},
    {"INT_LIT",    "[0-9]+"},

    {"INCREMENT",  "\\+\\+"},
    {"DECREMENT",  "--"},
    {"PLUS",       "\\+"},
    {"MINUS",      "-"},
    {"MULT",       "\\*"},
    {"DIV",        "/"},
    {"MOD",        "%"},

    {"EQ",         "=="},
    {"NE",         "!="},
    {"LE",         "<="},
    {"GE",         ">="},
    {"LT",         "<"},
    {"GT",         ">"},

    {"AND",        "&&"},
    {"OR",         "\\|\\|"},
    {"NOT",        "!"},

    {"BIT_AND",    "&"},
    {"BIT_OR",     "\\|"},
    {"BIT_XOR",    "\\^"},
    {"BIT_NOT",    "~"},

    {"ASSIGN",     "="},

    {"LPAREN",     "\\("},
    {"RPAREN",     "\\)"},
    {"LBRACE",     "\\{"},
    {"RBRACE",     "\\}"},
    {"LBRACKET",   "\\["},
    {"RBRACKET",   "\\]"},

    {"SEMICOLON",  ";"},
    {"COMMA",      ","},
    {"DOT",        "\\."},
    {"COLON",      ":"},
    {"QUESTION",   "\\?"},

    {"STRING_LIT", "\"([^\"\\\\]|\\\\.)*\""},
    {"CHAR_LIT",   "'([^'\\\\]|\\\\.)'"},

    {"WHITESPACE", "[[:space:]]+"}
};

#define TOKEN_COUNT (sizeof(TOKEN_SPEC) / sizeof(TOKEN_SPEC[0]))

/* ============================================================
   FUNÇÕES AUXILIARES
   ============================================================ */

int is_word_char(char c) {
    return isalnum((unsigned char)c) || c == '_';
}


/*
 * Verifica se um token de palavra possui limites de palavra,
 * equivalente ao \b usado no Python.
 */
int valid_word_boundary(
    const char *codigo,
    int inicio,
    int fim,
    int tamanho
) {
    if (inicio > 0 && is_word_char(codigo[inicio - 1])) {
        return 0;
    }

    if (fim < tamanho && is_word_char(codigo[fim])) {
        return 0;
    }

    return 1;
}


/*
 * Faz o escape necessário para produzir uma string JSON válida.
 */
void print_json_string(const char *str) {

    putchar('"');

    while (*str) {

        switch (*str) {

            case '"':
                printf("\\\"");
                break;

            case '\\':
                printf("\\\\");
                break;

            case '\n':
                printf("\\n");
                break;

            case '\r':
                printf("\\r");
                break;

            case '\t':
                printf("\\t");
                break;

            default:
                putchar(*str);
                break;
        }

        str++;
    }

    putchar('"');
}


/* ============================================================
   GET_ATTRIBUTE
   ============================================================ */

void get_attribute(
    const char *tipo,
    const char *valor,
    char *saida,
    size_t tamanho
) {

    if (strcmp(tipo, "IDENT") == 0) {

        snprintf(
            saida,
            tamanho,
            "\"%s\"",
            valor
        );

        return;
    }

    if (strcmp(tipo, "INT_LIT") == 0) {

        snprintf(
            saida,
            tamanho,
            "%d",
            atoi(valor)
        );

        return;
    }

    if (strcmp(tipo, "FLOAT_LIT") == 0) {

        snprintf(
            saida,
            tamanho,
            "%f",
            atof(valor)
        );

        return;
    }

    if (strcmp(tipo, "STRING_LIT") == 0) {

        int len = strlen(valor);

        if (len >= 2) {

            char temp[4096];

            strncpy(
                temp,
                valor + 1,
                len - 2
            );

            temp[len - 2] = '\0';

            snprintf(
                saida,
                tamanho,
                "\"%s\"",
                temp
            );
        }

        return;
    }

    if (strcmp(tipo, "CHAR_LIT") == 0) {

        int len = strlen(valor);

        if (len >= 2) {

            char temp[256];

            strncpy(
                temp,
                valor + 1,
                len - 2
            );

            temp[len - 2] = '\0';

            snprintf(
                saida,
                tamanho,
                "\"%s\"",
                temp
            );
        }

        return;
    }

    /*
     * Python retorna None para os demais tipos.
     */
    snprintf(
        saida,
        tamanho,
        "null"
    );
}


/* ============================================================
   LEXER
   ============================================================ */

void lexer(const char *codigo) {

    int pos = 0;
    int linha = 1;
    int coluna = 1;

    int tamanho_codigo = strlen(codigo);

    while (pos < tamanho_codigo) {

        int encontrou = 0;

        char tipo[100];
        char valor[4096];

        int tamanho_match = 0;

        /*
         * Tenta cada token na mesma ordem do TOKEN_SPEC
         */
        for (int i = 0; i < TOKEN_COUNT; i++) {

            regex_t regex;
            regmatch_t match;

            int resultado = regcomp(
                &regex,
                TOKEN_SPEC[i].pattern,
                REG_EXTENDED
            );

            if (resultado != 0) {
                continue;
            }

            /*
             * Cria uma string começando na posição atual.
             */
            const char *inicio = codigo + pos;

            resultado = regexec(
                &regex,
                inicio,
                1,
                &match,
                0
            );

            if (resultado == 0 && match.rm_so == 0) {

                int inicio_match = pos + match.rm_so;
                int fim_match = pos + match.rm_eo;

                /*
                 * Tokens de palavras precisam respeitar
                 * os limites de palavra.
                 */
                if (
                    strcmp(TOKEN_SPEC[i].name, "AUTO") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "BREAK") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "CASE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "CHAR") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "CONST") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "CONTINUE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "DEFAULT") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "DO") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "DOUBLE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "ELSE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "ENUM") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "EXTERN") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "FLOAT") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "FOR") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "GOTO") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "IF") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "INT") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "LONG") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "REGISTER") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "RETURN") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "SHORT") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "SIGNED") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "SIZEOF") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "STATIC") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "STRUCT") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "SWITCH") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "TYPEDEF") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "UNION") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "UNSIGNED") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "VOID") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "VOLATILE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "WHILE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "BOOL") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "TRUE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "FALSE") == 0 ||
                    strcmp(TOKEN_SPEC[i].name, "PRINT") == 0
                ) {

                    if (!valid_word_boundary(
                            codigo,
                            inicio_match,
                            fim_match,
                            tamanho_codigo
                        )) {

                        regfree(&regex);
                        continue;
                    }
                }

                tamanho_match = match.rm_eo;

                strncpy(
                    valor,
                    inicio,
                    tamanho_match
                );

                valor[tamanho_match] = '\0';

                strcpy(
                    tipo,
                    TOKEN_SPEC[i].name
                );

                encontrou = 1;

                regfree(&regex);

                break;
            }

            regfree(&regex);
        }

        /*
         * Nenhum token encontrado.
         */
        if (!encontrou) {

            fprintf(
                stderr,
                "Caractere inválido na linha %d e na coluna %d: '%c'\n",
                linha,
                coluna,
                codigo[pos]
            );

            exit(1);
        }

        int linha_inicio = linha;
        int coluna_inicio = coluna;

        pos += tamanho_match;

        /*
         * Atualização de linha e coluna.
         */
        for (int i = 0; i < tamanho_match; i++) {

            if (valor[i] == '\n') {

                linha++;
                coluna = 1;

            } else {

                coluna++;
            }
        }

        /*
         * Ignora espaços e comentários.
         */
        if (
            strcmp(tipo, "WHITESPACE") == 0 ||
            strcmp(tipo, "COMMENT") == 0 ||
            strcmp(tipo, "BLOCK_COMMENT") == 0
        ) {
            continue;
        }

        /*
         * Obtém o atributo.
         */
        char attribute[4096];

        get_attribute(
            tipo,
            valor,
            attribute,
            sizeof(attribute)
        );

        /*
         * Produz o mesmo formato JSON do Python.
         */
        printf("{");

        printf("\"token\":");
        print_json_string(tipo);

        printf(",");

        printf("\"lexeme\":");
        print_json_string(valor);

        printf(",");

        printf("\"attribute\":");
        printf("%s", attribute);

        printf(",");

        printf("\"line\":%d", linha_inicio);

        printf(",");

        printf("\"column\":%d", coluna_inicio);

        printf("}\n");
    }

    /*
     * EOF
     */
    printf("{");

    printf("\"token\":\"EOF\",");

    printf("\"lexeme\":\"\",");

    printf("\"attribute\":null,");

    printf("\"line\":%d,", linha);

    printf("\"column\":%d", coluna);

    printf("}\n");
}


/* ============================================================
   LER ARQUIVO
   ============================================================ */

char *ler_arquivo(const char *nome_arquivo) {

    FILE *arquivo = fopen(
        nome_arquivo,
        "rb"
    );

    if (arquivo == NULL) {

        fprintf(
            stderr,
            "Arquivo não encontrado: %s\n",
            nome_arquivo
        );

        exit(1);
    }

    /*
     * Vai para o final do arquivo para descobrir
     * seu tamanho.
     */
    fseek(
        arquivo,
        0,
        SEEK_END
    );

    long tamanho = ftell(arquivo);

    rewind(arquivo);

    /*
     * Reserva memória.
     */
    char *codigo = malloc(
        tamanho + 1
    );

    if (codigo == NULL) {

        fprintf(
            stderr,
            "Erro ao alocar memória.\n"
        );

        fclose(arquivo);

        exit(1);
    }

    /*
     * Lê o arquivo.
     */
    size_t lidos = fread(
        codigo,
        1,
        tamanho,
        arquivo
    );

    codigo[lidos] = '\0';

    fclose(arquivo);

    return codigo;
}


/* ============================================================
   MAIN
   ============================================================ */

int main(int argc, char *argv[]) {

    if (argc != 2) {

        printf(
            "Uso: scanner <arquivo.c>\n"
        );

        return 1;
    }

    const char *nome_arquivo = argv[1];

    char *codigo = ler_arquivo(
        nome_arquivo
    );

    lexer(codigo);

    free(codigo);

    return 0;
}